File: Visual Studio Product Team/Official Visual Studio 2008 VB.NET Samples/[VB]-Official Visual Studio 2008 VB.NET Samples/Simple Lambdas Sample (Visual Basic)/VB/SimpleLambdas/Samples.vb
'Copyright (C) Microsoft Corporation.  All rights reserved.

Imports System
Imports System.Collections
Imports System.Collections.Generic
Imports System.Linq

Public Class Samples
    Shared numbers As Integer() = New Integer() {5, 4, 1, 3, 9, 8, 6, 7, 2, 0}
    Shared strings As String() = New String() {"zero", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine"}

    Class Person
        Public Name As String
        Public Level As String
    End Class

    Shared persons As Person() = New Person() { _
        New Person With {.Name = "Jesper", .Level = 3}, _
        New Person With {.Name = "Lene", .Level = 3}, _
        New Person With {.Name = "Jonathan", .Level = 5}, _
        New Person With {.Name = "Sagiv", .Level = 3}, _
        New Person With {.Name = "Jacqueline", .Level = 3}, _
        New Person With {.Name = "Ellen", .Level = 3}, _
        New Person With {.Name = "Gustavo", .Level = 9}}


    Public Shared Sub Sample1()
        ' use Where() to filter out elements matching a particular condition       
        Dim fnums = numbers.Where(Function(n) n < 5)

        Console.WriteLine("Numbers < 5")
        For Each x As Integer In fnums
            Console.WriteLine(x)
        Next
    End Sub

    Public Shared Sub Sample2()
        ' use First() to find the one element matching a particular condition       
        Dim v As String = strings.First(Function(s) s(0) = "o"c)

        Console.WriteLine("string starting with 'o': {0}", v)
    End Sub

    Public Shared Sub Sample3()
        ' use Select() to convert each element into a new value
        Dim snums = numbers.Select(Function(n) strings(n))

        Console.WriteLine("Numbers")
        For Each s As String In snums
            Console.WriteLine(s)
        Next
    End Sub

    Public Shared Sub Sample4()
        ' use Anonymous Type constructors to construct multi-valued results on the fly
        Dim q = strings.Select(Function(s) New With {.Head = s.Substring(0, 1), .Tail = s.Substring(1)})
        For Each p In q
            Console.WriteLine("Head = {0}, Tail = {1}", p.Head, p.Tail)
        Next
    End Sub

    Public Shared Sub Sample5()
        ' Combine Select() and Where() to make a complete query
        Dim q = numbers.Where(Function(n) n < 5).Select(Function(n) strings(n))

        Console.WriteLine("Numbers < 5")
        For Each x In q
            Console.WriteLine(x)
        Next
    End Sub

    Public Shared Sub Sample6()
        ' use GroupBy() to construct group partitions out of similar elements
        Dim q = strings.GroupBy(Function(s) s(0)) ' <- group by first character of each string

        For Each g In q
            Console.WriteLine("Group: {0}", g.Key)
            For Each v As String In g
                Console.WriteLine(vbTab & "Value: {0}", v)
            Next
        Next
    End Sub

    Public Shared Sub Sample7()
        ' use GroupBy() and aggregates such as Count(), Min(), Max(), Sum(), Average() to compute values over a partition
        Dim q = strings.GroupBy(Function(s) s(0)).Select(Function(g) New With {.FirstChar = g.Key, .Count = g.Count()})

        For Each v In q
            Console.WriteLine("There are {0} string(s) starting with the letter {1}", v.Count, v.FirstChar)
        Next
    End Sub

    Public Shared Sub Sample8()
        ' use OrderBy()/OrderByDescending() to give order to your resulting sequence
        Dim q = strings.OrderBy(Function(s) s) ' order the strings by their name

        For Each s As String In q
            Console.WriteLine(s)
        Next
    End Sub

    Public Shared Sub Sample9()
        ' use ThenBy()/ThenByDescending() to provide additional ordering detail
        Dim q = persons.OrderBy(Function(p) p.Level).ThenBy(Function(p) p.Name)

        For Each p In q
            Console.WriteLine("{0}  {1}", p.Level, p.Name)
        Next
    End Sub

    Public Shared Sub Sample10()
        ' use query expressions to simplify
        Dim q = From p In persons _
                Order By p.Level, p.Name _
                Group p.Name By p.Level Into People = Group

        For Each g In q
            Console.WriteLine("Level: {0}", g.Level)
            For Each s In g.People
                Console.WriteLine("Person: {0}", s)
            Next
        Next
    End Sub
End Class

File: Visual Studio Product Team/Official Visual Studio 2010 Samples for C# 4.0/46872-Official Visual Studio 2010 Samples for C# 4.0/LINQ - Simple Lambdas/C#/Samples.cs
// Copyright © Microsoft Corporation.  All Rights Reserved.
// This code released under the terms of the 
// Microsoft Public License (MS-PL, http://opensource.org/licenses/ms-pl.html.)
//
//Copyright (C) Microsoft Corporation.  All rights reserved.

using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;

static class Samples {
    static int[] numbers = new [] { 5, 4, 1, 3, 9, 8, 6, 7, 2, 0 };
    static string[] strings = new [] { "zero", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine" };
    
    class Person {
      public string Name {get; set;}
      public int Level {get; set;}
    }
    
    static Person[] persons = new Person[] {
        new Person {Name="Jesper", Level=3},
        new Person {Name="Lene", Level=3},
        new Person {Name="Jonathan", Level=5},
        new Person {Name="Sagiv", Level=3},
        new Person {Name="Jacqueline", Level=3},
        new Person {Name="Ellen", Level=3},
        new Person {Name="Gustavo", Level=9}
        };
    
    public static void Sample1() {
        // use Where() to filter out elements matching a particular condition       
        var fnums = numbers.Where(n => n < 5);
    
        Console.WriteLine("Numbers < 5");
        foreach(int x in fnums) {
            Console.WriteLine(x);
        }
    }    

    public static void Sample2() {
        // use First() to find the one element matching a particular condition       
        string v = strings.First(s => s[0] == 'o');
    
        Console.WriteLine("string starting with 'o': {0}", v);
    }        
    
    public static void Sample3() {
        // use Select() to convert each element into a new value
        var snums = numbers.Select(n => strings[n]);
        
        Console.WriteLine("Numbers");
        foreach(string s in snums) {
            Console.WriteLine(s);
        }           
    }
    
    public static void Sample4()
    {
        // use Anonymous Type constructors to construct multi-valued results on the fly
        var q = strings.Select(s => new {Head = s.Substring(0,1), Tail = s.Substring(1)});
        foreach(var p in q) {
            Console.WriteLine("Head = {0}, Tail = {1}", p.Head, p.Tail);
        }
    }
    
    public static void Sample5() { 
        // Combine Select() and Where() to make a complete query
        var q = numbers.Where(n => n < 5).Select(n => strings[n]);
        
        Console.WriteLine("Numbers < 5");
        foreach(var x in q) {
            Console.WriteLine(x);
        }       
    }
    
    public static void Sample6() {
        // Sequence operators form first-class queries are not executed until you enumerate them.
        int i = 0;
        var q = numbers.Select(n => ++i);
        // Note, the local variable 'i' is not incremented until each element is evaluated (as a side-effect).
        foreach(var v in q) {
          Console.WriteLine("v = {0}, i = {1}", v, i);          
        }
        Console.WriteLine();
        
        // Methods like ToList() cause the query to be executed immediately, caching the results
        int i2 = 0;
        var q2 = numbers.Select(n => ++i2).ToList();
        // The local variable i2 has already been fully incremented before we iterate the results
        foreach(var v in q2) {
          Console.WriteLine("v = {0}, i2 = {1}", v, i2);
        }        
    }
    
    public static void Sample7() {
        // use GroupBy() to construct group partitions out of similar elements
        var q = strings.GroupBy(s => s[0]); // <- group by first character of each string
        
        foreach(var g in q) {
          Console.WriteLine("Group: {0}", g.Key);
          foreach(string v in g) {
            Console.WriteLine("\tValue: {0}", v);
          }
        }
    }
    
    public static void Sample8() {
        // use GroupBy() and aggregates such as Count(), Min(), Max(), Sum(), Average() to compute values over a partition
        var q = strings.GroupBy(s => s[0]).Select(g => new {FirstChar = g.Key, Count = g.Count()});
        
        foreach(var v in q) {
          Console.WriteLine("There are {0} string(s) starting with the letter {1}", v.Count, v.FirstChar);
        }
    }
    
    public static void Sample9() {
        // use OrderBy()/OrderByDescending() to give order to your resulting sequence
        var q = strings.OrderBy(s => s);  // order the strings by their name
        
        foreach(string s in q) {
          Console.WriteLine(s);
        }
    }
    
    public static void Sample9a() {  
        // use ThenBy()/ThenByDescending() to provide additional ordering detail
        var q = persons.OrderBy(p => p.Level).ThenBy(p => p.Name);
        
        foreach(var p in q) {
          Console.WriteLine("{0}  {1}", p.Level, p.Name);
        }
    }
    
    public static void Sample10() {
        // use query expressions to simplify
        var q = from p in persons 
                orderby p.Level, p.Name
                group p.Name by p.Level into g
                select new {Level = g.Key, Persons = g};
        
        foreach(var g in q) {
            Console.WriteLine("Level: {0}", g.Level);
            foreach(var p in g.Persons) {
                Console.WriteLine("Person: {0}", p);
            }
        }                
    }
}
